docs/internal/types.md

```diff
@@ -0,0 +1,287 @@
+
+## Primitive Types
+
+Verona treats numeric types specially in the compiler but their representation isn't special in the language.
+
+Verona's numeric types are singleton types (no fields) that have their machine representation based on their names.
+So, `I32` is a signed integer with 32-bit width (in LLVM's parlance, an `i32`) and so on.
+
+The full list of primitive numeric types are:
+ * Unsigned integers: `U8`, `U16`, `U32`, `U64` and `U128`
+ * Signed integers: `I8`, `I16`, `I32`, `I64` and `I128`
+ * Floating Point: `F32` and `F64`
+ * System types: `ULength` and `ILength` (machine-word sized integers)
+
+Note, integer literals have no defined type and their types (and representation) will be inferred by the compiler depending on the type of the variable being assigned to.
+
+For example:
+```ts
+// In this case, `42` has type `I32`
+let x : I32 = 42;
+
+// In this case, `42` has type `U64`
+let x : U64 = 42;
+
+// These are all compiler errors
+let x = 42; // No type to infer
+let x : I32 = 3.1415; // Type casts not allowed
+
+```
+
+## Type Unions
+
+Type unions means a variable can be any type in the list.
+
+For example:
+```ts
+foobar(a : (A | B | C))
+{
+  ...
+}
+
+main()
+{
+  let a : A;
+  let b : B;
+  let d : D;
+
+  // This is fine
+  foobar(a);
+  foobar(b);
+
+  // This is not
+  foobar(d); // ERROR
+}
+```
+
+The implementation of `foobar` will have to cope with all three types, either via a `match` statement or taking advantage of some behaviour that is common to all three (say, an interface they all implement).
+
+The compiler is free, however, to optimise the calls by creating two versions of `foobar` and replace the calls to the generic `foobar` for their specific ones.
+If the semantics are defined at compile time, this can simplifies a `match` statement into their specific matches and even inline the call altogether.
+
+Example:
+```ts
+// The generic one, if there are still dynamic cases
+foobar(a : (A | B | C)) { ... }
+
+// The specific specialisations used in defined cases
+foobar(a : A) { ... }
+foobar(b : B) { ... }
+```
+
+## Type Intersections
+
+Intersection means a type must all types involved.
+
+For example:
+```ts
+interface Serialisable { toString() : String; };
+
+main()
+{
+  let x : (A & Serialisable & mut) = ...;
+}
+```
+
+In the above example, the type of `x` must conform to both `A` and `Serialisable` guarantees (methods, fields) as well as the `mutable` capability.
+
+Every type in Verona must be an intersection of a type and a capability, so:
+```ts
+var x : (A & imm); // this is ok
+
+var y : B; // this is not
+```
+
+Type aliases can be used to make it easier to use types, for example:
+```ts
+class _String { ... };
+
+// We mostly want strings as values
+type String = (_String & imm);
+
+// But we may want to change it
+type MutableString = (_String & mut);
```

Comment:
Hmm, I'm not sure what this means.  Is `String` a public thing, even though `_String` is package-private?  What happens if you have a public type alias of a private type?

Reply:
I haven't touched public/private in this document, perhaps we should. Are public fields those prepended with `public`? Like `public class Foo {}` or `public type Foo = Bar`? I'm still unsure how we can expose some fields outside the module.

For example:
```
module 1 {
  class Private {
    public var foo; // is this legal?
  };
  public class Public {
    private var bar; // or this?
  };
```

If only types can public, then one way to expose would be to create a facade:
```
module 1 {
  public interface Public {
    create() : Public {
      var self = Private::create();
      return self;
    }
  };
  class Private : Public { ... };
}
```

but this is convoluted. It will depend on how we control visibility: type level, field/method level, etc.

Reply:
> I haven't touched public/private in this document, perhaps we should. Are public fields those prepended with public? Like public class Foo {} or public type Foo = Bar? I'm still unsure how we can expose some fields outside the module.

Either with `public` / `private` or with a naming convention (`_`-prefix = private to the package, anything else is public).  If a thing is public, it can be named outside the current package.  I think there are still a few corner cases though (more below).

> For example:

All of this is legal.

> If only types can public

Types, fields, and methods, can be public or (package) private.

```
module m1 // fake syntax, modules are classes in different places
{
  class Foo
  {
     var x : U64;
     var _privateField : U64;
     method(self : Foo) {...}
     _privateMethod(self :Foo) {...}
  }
}
// In another module, having imported m1 as m1
var f : Foo = ...; // This is fine
f.x  =12; //fine.
f._privateField = 12; //Nope
f.method(); //fine
f._privateMethod(); //Nope
```

This gets more interesting with interfaces:

```
module m1 // same one as above
{
  interface I1
  {
     var x : U64;
     var _privateField : U64;
  }
  getAnI1() : I1
  {
     var ret : Foo = ...;
     return ret;
  }
}
module m2
{
  // Assume m1 imported as m1
  var i1 = m1.getAnI1();
  i1.x = 42; // Yup
  i1._privateField = 12; // Maybe?  
  // Okay, what about this (not in m1)?
  interface I2
  {
     var _privateField : U64;
  }
  // Is this allowed?
  var i1 : I2 = i1;
}
```

I like a ClassBox-like approach, where all private selectors and type names are logically mangled with your local module ID, so `_privateField` as named in `m1` and in the package containing `I2` are different selectors and the last cast is not permitted.  It's less clear to me what this means for the `i1._privateField` access, it might simply mean that this access is statically not permitted because this is logically  the same as saying `i1._privateField@m2` and `I1` defines a field called `_privateField@m1`, not a field called `_privateField@m2` and so you may not access this.

There's one final case:

```
module m3
{
  interface I4 {...};
  class _Cls { ... } // Implements I4
  type Cls = _Cls;
  getI4() : I4 { /* return a _Cls instance */ }
}
//Elsewhere:
var i4=  m3::getI4();
match i4 { cls : m4::_Cls => ...} // Nope, you may not name m4::_Cls from here.
match i4 { cls : m4::Cls => ...} // Maybe?  Did the type alias make Cls public?
```

Reply:
I think this deserves more discussions, but I've added a new line to the example to mean this is only about the same module, so there are no visibility concerns.

We should address visibility in another section or perhaps another document.

Reply:
I'd prefer that you didn't name the concrete type with an underscore prefix (or you named everything with an underscore prefix) because then there's no ambiguity.

Reply:
Fair enough, renamed.